phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -1571,15 +1576,28 @@ public GetVersionResponse call(MetaDataService instance)
                 }
             }
         }
-        if (isIncompatible) {
-            buf.setLength(buf.length()-1);
-            throw new SQLExceptionInfo.Builder(SQLExceptionCode.OUTDATED_JARS).setMessage(buf.toString()).build().buildException();
+        if (!(errorCode.toString().equals(new StringBuilder().toString()))) {
+            if (Integer.parseInt(errorCode.toString()) == SQLExceptionCode.OUTDATED_JARS.getErrorCode()) {
+                errorMessage.setLength(errorMessage.length()-1);
+                throw new SQLExceptionInfo.Builder(SQLExceptionCode.OUTDATED_JARS).setMessage(errorMessage.toString()).build().buildException();
+            } else {
+                throw new SQLExceptionInfo.Builder(SQLExceptionCode.INCOMPATIBLE_CLIENT_SERVER_JAR).setMessage(errorMessage.toString()).build().buildException();
+            }
         }
+
         if (systemCatalogTimestamp < MIN_SYSTEM_TABLE_TIMESTAMP) {
             throw new UpgradeRequiredException(systemCatalogTimestamp);
         }
     }
 
+    private String displayClientandServerVersions(long serverJarVersion) {
+        String clientVersion = PHOENIX_MAJOR_VERSION + "." + PHOENIX_MINOR_VERSION + "." + PHOENIX_PATCH_NUMBER;
```

Comment:
This is already present , you can probably just use it : https://github.com/apache/phoenix/blob/master/phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataProtocol.java#L151